### DISCUSSION TITLE
A base type for annotations?

### DISCUSSION THREAD

**Discussion by Dreamsorcerer** [python/typing#1543] [General]
I don't think this is possible currently, but it would be useful for Generics etc. if there was some way to specify `type[T]` for any valid annotation, to be used at runtime.

For example, we have `aiohttp.web.AppKey` defined like:
```
class AppKey(Generic[T]):
    def __init__(self, name: str, t: type[T]):
```
expected to be used like `AppKey("foo", int)` to define the type for the key (mostly imitating the design of TypeVar).

The problem is that a lot of valid typing annotations are not actual 'types' and produce a typing error (e.g. `AppKey(..., Callable)`). There is probably some issue with making all of these compatible with `type`, but maybe we could introduce a new thing for types of typing constructs? Then we can annotate it like `type[T] | annotation_type[T]` or something.

**Comment by Gobot1234** [comment]
This was something I wanted to fix in a future pep for reified types

**Comment by JelleZijlstra** [comment]
This sounds like you want TypeForm: https://github.com/python/mypy/issues/9773. That proposal hasn't progressed much, though.
